src/main/scala/dev/kamu/cli/commands/PullCommand.scala

```diff
@@ -20,15 +20,44 @@ class PullCommand(
   private val logger = LogManager.getLogger(getClass.getName)
 
   def run(): Unit = {
-    val datasets =
-      if (all)
-        metadataRepository.getAllDatasets()
-      else
-        datasetIDs
-          .map(DatasetID)
-          .map(metadataRepository.getDataset)
-
-    val numUpdated = datasets
+    val s = metadataRepository.getAllDatasetIDs()
+    val datasets: IndexedSeq[DatasetID] = s.toIndexedSeq
+    val datasetIntID = s.zipWithIndex.toMap
+    val idToInt: DatasetID => Int = d => datasetIntID(d)
+    val intToId: Int => DatasetID = d => datasets(d)
+    val idToDataset: DatasetID => Dataset = d => metadataRepository.getDataset(d)
+
+    val scheduled: Array[Boolean] = new Array(datasets.length)
+    for (i <- scheduled.indices) scheduled.update(i, false)
+    var toSchedule: List[Int] = Nil
+    var plan: Seq[Int] = Nil
+    def isScheduled(d: Int): Boolean = scheduled(d)
+
+    if (all) toSchedule = datasets.indices.toList.reverse
+    else toSchedule = datasetIDs.map(DatasetID).toList.map(idToInt)
+
+    logger.info(s"Your pull request:")
+    for (d <- toSchedule) logger.info(s"  ${intToId(d).toString}")
+
+    while (Nil != toSchedule) {
+      val x: Int = toSchedule.head
+      val xs: List[Int] = toSchedule.tail
+      val toResolve: Seq[Int] = idToDataset(intToId(x))
+        .dependsOn.map(idToInt).filter(isScheduled _ andThen (!_))
+      if (0 == toResolve.length) {
+        if (!isScheduled(x)) {
+          plan = plan :+ x
+          scheduled.update(x, true)
+        }
+        toSchedule = xs
+      } else toSchedule = toResolve.foldRight(toSchedule)(_ :: _)
+    }
+
+    logger.info(s"This is what I am about to pull in order:")
```

Comment:
Currently everything that is logged at INFO and above is displayed to the user as output in stderr, so be mindful of this. When user pulls datasets they probably only care that this happened and trust us to figure out correct order of things, so I think this should be logged at DEBUG.

Also I prefer a "Doing BLAH" style of logging (imperative style??) as it's least verbose, e.g. "Pulling datasets in following order: ..."